File: akka-samples/java-samples/src/main/java/actors/state/StatePrinter.java
package actors.state;

import akka.actor.typed.ActorSystem;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;

class StatePrinter extends AbstractBehavior<StatePrinter.PrintMessage> {
    static final class PrintMessage {
        public final String text;

        public PrintMessage(String text) {
            this.text = text;
        }
    }

    private final int maximumMessages = 3;
    private int messageCounter = 0;

    @Override
    public Receive<PrintMessage> createReceive() {
        return newReceiveBuilder().onMessage(PrintMessage.class, (message) -> {
            System.out.println(message.text);

            int messagesTillShutdown = maximumMessages - ++messageCounter;
            System.out.printf("%s messages left before shutdown\r\n", messagesTillShutdown);

            if (messagesTillShutdown == 0) {
                System.out.println("Shutting down!");
                return Behaviors.stopped();
            }

            return Behaviors.same();
        }).build();
    }

    public StatePrinter(ActorContext<PrintMessage> context) {
        super(context);
    }

    public static Behavior<PrintMessage> create() {
        return Behaviors.setup(StatePrinter::new);
    }

    public static void main(String[] args) {
        final ActorSystem<PrintMessage> system = ActorSystem.create(StatePrinter.create(), "printer");
        system.tell(new PrintMessage("Hello there!"));
        system.tell(new PrintMessage("Oh, Hi there!"));
        system.tell(new PrintMessage("Oh, Hi Mark!"));
    }
}

File: akka-samples/scala-samples/src/main/scala/actors/state/StatePrinter.scala
package actors.state

import akka.actor.typed.scaladsl.Behaviors
import akka.actor.typed.{ActorSystem, Behavior}
import actors.state.Printer.PrintMessage

object StatePrinter {
  def main(args: Array[String]): Unit = {
    val system = ActorSystem[PrintMessage](Printer(), "state-printer")
    system ! PrintMessage("Hello there!")
    system ! PrintMessage("Oh, Hi there!")
    system ! PrintMessage("Oh, Hi Mark!")
  }
}

private object Printer {
  final case class PrintMessage(text: String)

  private val maxMessages: Int = 3
  private var messageCounter: Int = 0

  def apply(): Behavior[PrintMessage] =
    Behaviors.receive[PrintMessage] { (context, message) =>
      println(message.text)

      messageCounter = messageCounter + 1
      val messagesTillShutdown = maxMessages - messageCounter
      println(s"$messagesTillShutdown messages left before shutdown\r\n")

      if (messagesTillShutdown == 0) {
        println("Shutting down!")
        Behaviors.stopped
      }

      Behaviors.same
    }
}